pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+	if err != nil {
+		klog.Errorf("Reconciler Failed to list Shares: %v", err)
+	}
+	instances, err := recon.cloud.File.ListMultishareInstances(context.TODO(), &file.ListFilter{Project: recon.cloud.Project, Location: "-"})
+	if err != nil {
+		klog.Errorf("Reconciler Failed to list Instances: %v", err)
+	}
+	klog.V(6).Infof("found %d shares and %d instances", len(shares), len(instances))
+	// ops, err := recon.cloud.File.ListOps(context.TODO(), &file.ListFilter{Project: recon.cloud.Project, Location: "-"})
+	// if err != nil {
+	// 	klog.Errorf("Reconciler Failed to list ops: %v", err)
+	// }
+	instances, shares, instanceShares := recon.managedInstanceAndShare(instances, shares)
+
+	shareInfoMap := recon.createAndUpdateShareInfos(shares)
+
+	shareInfoList, err := recon.shareLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list ShareInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("listed out %d shareInfo objects", len(shareInfoList))
+	for _, shareInfo := range shareInfoList {
+		shareName := shareInfo.Name
+		if _, ok := shareInfoMap[shareName]; !ok {
+			shareInfoMap[shareName] = shareInfo
+		}
+	}
+
+	instanceInfoMap := recon.createAndUpdateInstanceInfos(instances, instanceShares)
+
+	instanceInfoList, err := recon.instanceLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list InstanceInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("listed out %d instanceInfo objects", len(instanceInfoList))
+	for _, instanceInfo := range instanceInfoList {
+		instanceURI := util.IINameToInstanceURI(instanceInfo.Name)
+		if _, ok := instanceInfoMap[instanceURI]; !ok {
+			instanceInfo, err := recon.maybeRemoveInstanceInfoFinalizer(instanceInfo)
+			if err != nil {
+				klog.Errorf("error removing Finalizer from InstanceInfo %q: %v", instanceInfo.Name, err)
+				continue
+			}
+			if instanceInfo != nil {
+				instanceInfoMap[instanceURI] = instanceInfo
+			}
+		}
+	}
+
+	recon.assignSharesToInstances(shareInfoMap, instanceInfoMap)
+
+	klog.Infof("reconciliation round finished after %v", time.Since(startTime))
+}
+
+func (recon *multishareReconciler) assignSharesToInstances(shareInfos map[string]*v1alpha1.ShareInfo, instanceInfos map[string]*v1alpha1.InstanceInfo) {
+	for _, instanceInfo := range instanceInfos {
+		if instanceInfo.Status == nil {
+			klog.Infof("instance %q has Status nil", instanceInfo.Name)
+			continue
+		}
+
+		instanceInfoClone := instanceInfo.DeepCopy()
+		updated := false
+
+		instanceURI := util.IINameToInstanceURI(instanceInfo.Name)
+		for _, shareName := range instanceInfo.Status.ShareNames {
+			shareInfo, ok := shareInfos[shareName]
+			if !ok {
+				klog.Errorf("share %q is assigned to instance %q but shareInfo does not exist", shareName, instanceURI)
+				continue
+			}
+			if shareInfo.Status == nil || shareInfo.Status.InstanceHandle == "" {
+				shareInfo, err := recon.assignShareInfo(shareInfo, instanceURI)
```

Comment:
should we skip this assignment if a share is marked for delete.

Reply:
yes. done.